{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Secretory Pathway Features Retrieval\n",
    "This notebook retrieves information diferent identifiers from the NIH database. and adds it to the **\"Secretory Pathway Recon\" Google Sheet**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load packages and define datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "import Request_Utilis\n",
    "from google_sheet import GoogleSheet\n",
    "\n",
    "Entrez.email = \"a.antonakoudis@sartorius.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### ----- Generate datasets from Google Sheet ----- #####\n",
    "\n",
    "#Credential file\n",
    "KEY_FILE_PATH = 'credentials.json'\n",
    "\n",
    "#CHO Network Reconstruction + Recon3D_v3 Google Sheet ID\n",
    "SPREADSHEET_ID = '1DaAdZlvMYDqb7g31I5dw-ZCZH52Xj_W3FnQMFUzqmiQ'\n",
    "\n",
    "# Initialize the GoogleSheet object\n",
    "gsheet_file = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "\n",
    "# Read data from the Google Sheet\n",
    "sec_recon_sheet = 'SecRecon'\n",
    "sec_recon = gsheet_file.read_google_sheet(sec_recon_sheet)\n",
    "# Create a copy of the dataset\n",
    "sec_recon_dc = sec_recon.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Retrieve Human CHO and Mouse Entrez IDs\n",
    "Here we use the fucntion get_entrez_id from the **Request Utilis** module to fetch the Entrez IDs for Human and then use this as input to retrieve information for CHO and Mouse."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1 Human Entriz ID\n",
    "Here we used the Human Entrez IDs from the step before to get the CHO and Mouse orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Update Human Entrez IDs\n",
    "for i,row in sec_recon_dc.iterrows():\n",
    "    if pd.isnull(row['HUMAN ENTREZID']) or row['HUMAN ENTREZID'] == '':\n",
    "        human_entrez = Request_Utilis.get_entrez_id(row['GENE SYMBOL'])\n",
    "        sec_recon_dc.at[i, 'HUMAN ENTREZID'] = human_entrez\n",
    "\n",
    "if not sec_recon_dc.equals(sec_recon):\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(\"Google Sheet updated.\")\n",
    "else:\n",
    "    print('Human Entrez IDs are up-to-date')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 CHO Entrez IDs from other databases\n",
    "Before running the **get_gene_id** function on CHO genes, we first populate some of the CHO genes with a mapping of orthologs based on our own dataset comprised from different databases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map Human IDs to CHO IDs from the \"cho2human_mapping\" dataset\n",
    "\n",
    "cho2human_mapping = pd.read_csv(\"../Orthologs/cho2human_mapping.tsv\", sep='\\t')\n",
    "cho2human_mapping2 = pd.read_excel(\"../Orthologs/orthologs.xlsx\", index_col=0)\n",
    "cho2human_mapping2['Human GeneID'] = pd.to_numeric(cho2human_mapping2['Human GeneID'], errors='coerce')\n",
    "cho2human_mapping2['Human GeneID'] = cho2human_mapping2['Human GeneID'].astype('Int64')\n",
    "\n",
    "cho_id_lookup = dict(zip(cho2human_mapping['HUMAN_ID'], cho2human_mapping['CHO_ID'])) #convert to dict for mapping\n",
    "cho_id_lookup2 = dict(zip(cho2human_mapping2['Human GeneID'], cho2human_mapping2['CHO GeneID'])) #convert to dict for mapping\n",
    "\n",
    "for index, row in sec_recon_dc.iterrows():\n",
    "    if pd.isna(row['CHO ENTREZID']) or row['CHO ENTREZID'] == '':\n",
    "        try:\n",
    "            human_id = int(row['HUMAN ENTREZID'])\n",
    "            cho_id = cho_id_lookup.get(human_id)\n",
    "            if cho_id is not None:\n",
    "                sec_recon_dc.at[index, 'CHO ENTREZID'] = cho_id\n",
    "            else:\n",
    "                try:\n",
    "                    cho_id = cho_id_lookup2.get(human_id)\n",
    "                    if cho_id is not None:\n",
    "                        sec_recon_dc.at[index, 'CHO ENTREZID'] = cho_id\n",
    "                except ValueError:\n",
    "                    print(f'{human_id} is not a valid Human Entrez ID')      \n",
    "        except ValueError:\n",
    "            print(f'{human_id} is not a valid Human Entrez ID')\n",
    "            continue        \n",
    "\n",
    "if not sec_recon_dc.equals(sec_recon):\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(\"Google Sheet updated on CHO Entrez IDs from cho2human dataset\")\n",
    "else:\n",
    "    print('CHO Entrez IDs from \"cho2human_mapping\" dataset are up-to-date')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 CHO and Mouse Entrez IDs \n",
    "Finally we run the **get_gene_ids** function to retrieve CHO and Mouse Entrez IDs by mapping the orthologs using the Human Entrez IDs as input."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## -- CHO Entrez IDs -- ##\n",
    "\n",
    "for index, row in sec_recon_dc.iterrows():\n",
    "    if pd.isna(row['CHO ENTREZID']) or row['CHO ENTREZID'] == '':\n",
    "        human_id = row['HUMAN ENTREZID']\n",
    "        cho_ortholog_EntrezID = Request_Utilis.get_gene_ids(human_id, '10029')\n",
    "        if cho_ortholog_EntrezID is not None:\n",
    "            sec_recon_dc.at[index, 'CHO ENTREZID'] = cho_ortholog_EntrezID\n",
    "            \n",
    "if not sec_recon_dc.equals(sec_recon):\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(\"Google Sheet updated on CHO Entrez IDs from NIH database\")\n",
    "else:\n",
    "    print('CHO Entrez IDs from NIH database are up-to-date')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## -- Mouse Entrez IDs -- ##\n",
    "\n",
    "loop_counter = 0\n",
    "update_threshold = 50\n",
    "\n",
    "for index, row in sec_recon_dc.iterrows():\n",
    "    if pd.isna(row['MOUSE ENTREZID']) or row['MOUSE ENTREZID'] == '':\n",
    "        human_id = row['HUMAN ENTREZID']\n",
    "        mouse_ortholog_EntrezID = Request_Utilis.get_gene_ids(human_id, '10090')\n",
    "        if mouse_ortholog_EntrezID is not None:\n",
    "            sec_recon_dc.at[index, 'MOUSE ENTREZID'] = mouse_ortholog_EntrezID\n",
    "            loop_counter += 1\n",
    "\n",
    "        if loop_counter >= update_threshold:\n",
    "            if not sec_recon_dc.equals(sec_recon):\n",
    "                gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "                print(f\"Google Sheet updated on Mouse Entrez IDs from NIH database after {loop_counter} updates\")\n",
    "            else:\n",
    "                print('Mouse Entrez IDs from NIH database are up-to-date')\n",
    "            loop_counter = 0\n",
    "\n",
    "# Check if there are any remaining updates after exiting the loop\n",
    "if loop_counter > 0 and not sec_recon_dc.equals(sec_recon):\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(f\"Google Sheet updated on Mouse Entrez IDs from NIH database after {loop_counter} updates\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Ensembl IDs\n",
    "In this section we retrieve Ensembl IDs fron NIH database using the **Gene_Info_from_EntrezID** function from the Request Utilis module. Secondarily, we retrieve extra information from other identifiers to fill missing data in our dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 Human Ensembl IDs and Extra Identifiers\n",
    "Here we retrieve the Human Ensembl IDs and Gene Alises and Gene Names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collect missing information from NIH database\n",
    "\n",
    "updates = []\n",
    "for i, gene in sec_recon_dc.iterrows():\n",
    "    human_entrezID = gene['HUMAN ENTREZID']\n",
    "    gene_symbol = gene['GENE SYMBOL']\n",
    "    if gene['ALIAS'] == '' or gene['GENENAME'] == '' or gene['HUMAN ENSEMBL'] == '':\n",
    "        print(gene_symbol)\n",
    "        try:\n",
    "            org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products = Request_Utilis.Gene_Info_from_EntrezID(human_entrezID)\n",
    "            updates.append((i, gene_synonyms, gene_name, gene_ensemble))\n",
    "        except ValueError:\n",
    "            print(f'No valid Entrez ID for gene {gene_symbol}')\n",
    "\n",
    "# Apply the updates outside the loop\n",
    "for i, gene_synonyms, gene_name, gene_ensemble in updates:\n",
    "    sec_recon_dc.at[i, 'ALIAS'] = gene_synonyms\n",
    "    sec_recon_dc.at[i, 'GENENAME'] = gene_name\n",
    "    sec_recon_dc.at[i, 'HUMAN ENSEMBL'] = gene_ensemble\n",
    "    \n",
    "sec_recon_dc['ALIAS'] = sec_recon_dc['ALIAS'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)\n",
    "sec_recon_dc['GENENAME'] = sec_recon_dc['GENENAME'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)\n",
    "sec_recon_dc['HUMAN ENSEMBL'] = sec_recon_dc['HUMAN ENSEMBL'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)\n",
    " \n",
    "    \n",
    "if not sec_recon_dc.equals(sec_recon):\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(\"Google Sheet updated.\")\n",
    "else:\n",
    "    print('Human identifiers are up-to-date')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 CHO and Mouse Ensembl IDs and Gene Symbols\n",
    "Using the same functionw we retrieve Ensembl IDs and Gene Symbols for CHO and Mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## -- CHO Ensembl IDs and Gene Symbol -- ##\n",
    "\n",
    "loop_counter = 0\n",
    "update_threshold = 50\n",
    "\n",
    "# Collect missing information for CHO identifiers\n",
    "for i, gene in sec_recon_dc.iterrows():\n",
    "    cho_entrezID = str(gene['CHO ENTREZID'])\n",
    "    if cho_entrezID != '':\n",
    "        if (pd.isna(gene['CHO ENSEMBL']) or gene['CHO ENSEMBL'] == '') or (pd.isna(gene['CHO GENE SYMBOL']) or gene['CHO GENE SYMBOL'] == ''):\n",
    "            try:\n",
    "                org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products = Request_Utilis.Gene_Info_from_EntrezID(cho_entrezID)\n",
    "                if (pd.isna(gene['CHO GENE SYMBOL']) or gene['CHO GENE SYMBOL'] == ''):\n",
    "                    sec_recon_dc.at[i, 'CHO GENE SYMBOL'] = gene_symbol\n",
    "                if (pd.isna(gene['CHO ENSEMBL']) or gene['CHO ENSEMBL'] == ''):\n",
    "                    sec_recon_dc.at[i, 'CHO ENSEMBL'] = gene_ensemble\n",
    "            except ValueError:\n",
    "                print(f'No valid Entrez ID for gene {gene_symbol}')\n",
    "            loop_counter += 1\n",
    "\n",
    "            if loop_counter >= update_threshold:\n",
    "                if not sec_recon_dc.equals(sec_recon):\n",
    "                    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "                    print(f\"Google Sheet updated on CHO Ensembl IDs after {loop_counter} updates\")\n",
    "                else:\n",
    "                    print('CHO Ensembl IDs are up-to-date')\n",
    "                loop_counter = 0\n",
    "\n",
    "# Check if there are any remaining updates after exiting the loop\n",
    "if loop_counter > 0 and not sec_recon_dc.equals(sec_recon):\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(f\"Google Sheet updated on CHO Ensembl IDs after {loop_counter} updates\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## -- Mouse Ensembl IDs and Gene Symbol-- ##\n",
    "\n",
    "loop_counter = 0\n",
    "update_threshold = 50\n",
    "\n",
    "# Collect missing information for CHO identifiers\n",
    "for i, gene in sec_recon_dc.iterrows():\n",
    "    mouse_entrezID = str(gene['MOUSE ENTREZID'])\n",
    "    if mouse_entrezID != '':\n",
    "        if (pd.isna(gene['MOUSE ENSEMBL']) or gene['MOUSE ENSEMBL'] == '') or (pd.isna(gene['MOUSE GENE SYMBOL']) or gene['MOUSE GENE SYMBOL'] == ''):\n",
    "            try:\n",
    "                org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products = Request_Utilis.Gene_Info_from_EntrezID(mouse_entrezID)\n",
    "                if (pd.isna(gene['MOUSE GENE SYMBOL']) or gene['MOUSE GENE SYMBOL'] == ''):\n",
    "                    sec_recon_dc.at[i, 'MOUSE GENE SYMBOL'] = gene_symbol\n",
    "                if (pd.isna(gene['MOUSE ENSEMBL']) or gene['MOUSE ENSEMBL'] == ''):\n",
    "                    sec_recon_dc.at[i, 'MOUSE ENSEMBL'] = gene_ensemble\n",
    "            except ValueError:\n",
    "                print(f'No valid Entrez ID for gene {gene_symbol}')\n",
    "            loop_counter += 1\n",
    "\n",
    "            if loop_counter >= update_threshold:\n",
    "                if not sec_recon_dc.equals(sec_recon):\n",
    "                    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "                    print(f\"Google Sheet updated on Mouse Ensembl IDs after {loop_counter} updates\")\n",
    "                else:\n",
    "                    print('Mouse Ensembl IDs are up-to-date')\n",
    "                loop_counter = 0\n",
    "\n",
    "# Check if there are any remaining updates after exiting the loop\n",
    "if loop_counter > 0 and not sec_recon_dc.equals(sec_recon):\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(f\"Google Sheet updated on Mouse Ensembl IDs after {loop_counter} updates\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Uniprot IDs\n",
    "In this section we retrieve all the Uniprot IDs linked to each gene Entrez ID from NIH database, using the **Gene_Info_from_EntrezID** function from the Request Utilis module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## -- Human Uniprot IDs -- ##\n",
    "\n",
    "loop_counter = 0\n",
    "update_threshold = 50\n",
    "\n",
    "# Collect missing information for CHO identifiers\n",
    "for i, gene in sec_recon_dc.iterrows():\n",
    "    human_entrezID = str(gene['HUMAN ENTREZID'])\n",
    "    if human_entrezID != '':\n",
    "        if (pd.isna(gene['HUMAN UNIPROT']) or gene['HUMAN UNIPROT'] == ''):\n",
    "            try:\n",
    "                org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products = Request_Utilis.Gene_Info_from_EntrezID(human_entrezID)\n",
    "                unique_uniprotids = list(set([item for sublist in [x[2] for x in gene_products] for item in sublist]))\n",
    "                sec_recon_dc.at[i, 'HUMAN UNIPROT'] = unique_uniprotids\n",
    "                print(loop_counter+1, gene_symbol, human_entrezID, unique_uniprotids)\n",
    "            except ValueError:\n",
    "                print(f'No valid Entrez ID for gene {gene_symbol}')\n",
    "            loop_counter += 1\n",
    "\n",
    "            if loop_counter >= update_threshold:\n",
    "                if not sec_recon_dc.equals(sec_recon):\n",
    "                    sec_recon_dc['HUMAN UNIPROT'] = sec_recon_dc['HUMAN UNIPROT'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)\n",
    "                    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "                    print(f\"Google Sheet updated on Human Uniprot IDs after {loop_counter} updates\")\n",
    "                else:\n",
    "                    print('HUMAN Uniprot IDs are up-to-date')\n",
    "                loop_counter = 0\n",
    "\n",
    "# Check if there are any remaining updates after exiting the loop\n",
    "if loop_counter > 0 and not sec_recon_dc.equals(sec_recon):\n",
    "    sec_recon_dc['HUMAN UNIPROT'] = sec_recon_dc['HUMAN UNIPROT'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(f\"Google Sheet updated on Human Uniprot IDs after {loop_counter} updates\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## -- CHO Uniprot IDs -- ##\n",
    "\n",
    "loop_counter = 0\n",
    "update_threshold = 50\n",
    "\n",
    "# Collect missing information for CHO identifiers\n",
    "for i, gene in sec_recon_dc.iterrows():\n",
    "    cho_entrezID = str(gene['CHO ENTREZID'])\n",
    "    if cho_entrezID != '':\n",
    "        if (pd.isna(gene['CHO UNIPROT']) or gene['CHO UNIPROT'] == ''):\n",
    "            try:\n",
    "                org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products = Request_Utilis.Gene_Info_from_EntrezID(cho_entrezID)\n",
    "                unique_uniprotids = list(set([item for sublist in [x[2] for x in gene_products] for item in sublist]))\n",
    "                sec_recon_dc.at[i, 'CHO UNIPROT'] = unique_uniprotids\n",
    "                print(loop_counter+1, gene_symbol, cho_entrezID, unique_uniprotids)\n",
    "            except ValueError:\n",
    "                print(f'No valid Entrez ID for gene {gene_symbol}')\n",
    "            loop_counter += 1\n",
    "\n",
    "            if loop_counter >= update_threshold:\n",
    "                if not sec_recon_dc.equals(sec_recon):\n",
    "                    sec_recon_dc['CHO UNIPROT'] = sec_recon_dc['CHO UNIPROT'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)\n",
    "                    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "                    print(f\"Google Sheet updated on CHO Uniprot IDs after {loop_counter} updates\")\n",
    "                else:\n",
    "                    print('CHO Uniprot IDs are up-to-date')\n",
    "                loop_counter = 0\n",
    "\n",
    "# Check if there are any remaining updates after exiting the loop\n",
    "if loop_counter > 0 and not sec_recon_dc.equals(sec_recon):\n",
    "    sec_recon_dc['CHO UNIPROT'] = sec_recon_dc['CHO UNIPROT'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(f\"Google Sheet updated on CHO Uniprot IDs after {loop_counter} updates\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## -- Mouse Uniprot IDs -- ##\n",
    "\n",
    "loop_counter = 0\n",
    "update_threshold = 50\n",
    "\n",
    "# Collect missing information for CHO identifiers\n",
    "for i, gene in sec_recon_dc.iterrows():\n",
    "    mouse_entrezID = str(gene['MOUSE ENTREZID'])\n",
    "    if mouse_entrezID != '':\n",
    "        if (pd.isna(gene['MOUSE UNIPROT']) or gene['MOUSE UNIPROT'] == ''):\n",
    "            try:\n",
    "                org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products = Request_Utilis.Gene_Info_from_EntrezID(mouse_entrezID)\n",
    "                unique_uniprotids = list(set([item for sublist in [x[2] for x in gene_products] for item in sublist]))\n",
    "                sec_recon_dc.at[i, 'MOUSE UNIPROT'] = unique_uniprotids\n",
    "                print(loop_counter+1, gene_symbol, mouse_entrezID, unique_uniprotids)\n",
    "            except ValueError:\n",
    "                print(f'No valid Entrez ID for gene {gene_symbol}')\n",
    "            loop_counter += 1\n",
    "\n",
    "            if loop_counter >= update_threshold:\n",
    "                if not sec_recon_dc.equals(sec_recon):\n",
    "                    sec_recon_dc['MOUSE UNIPROT'] = sec_recon_dc['MOUSE UNIPROT'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)\n",
    "                    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "                    print(f\"Google Sheet updated on Mouse Uniprot IDs after {loop_counter} updates\")\n",
    "                else:\n",
    "                    print('Mouse Uniprot IDs are up-to-date')\n",
    "                loop_counter = 0\n",
    "\n",
    "# Check if there are any remaining updates after exiting the loop\n",
    "if loop_counter > 0 and not sec_recon_dc.equals(sec_recon):\n",
    "    sec_recon_dc['MOUSE UNIPROT'] = sec_recon_dc['MOUSE UNIPROT'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(f\"Google Sheet updated on Mouse Uniprot IDs after {loop_counter} updates\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Subcellular Localization\n",
    "The subcellular localization is divided into two parts. First, we map the subcellular localization to all the genes from the data provided in the paper \"[Global organelle profiling reveals subcellular localization and remodeling at proteome scale](https://www.biorxiv.org/content/10.1101/2023.12.18.572249v1)\". Then, we use the **get_subcellular_localization** from the Request Utilis module to retrieve the subcellular localization of each gene using as input the Uniprot IDs retrieved previously."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate \"subcell_dict\" for direct mapping into our dataset\n",
    "subcell = pd.read_csv(\"../Input/subcellular_localization.csv\")\n",
    "subcell_dict = dict(zip(subcell['Gene_name_canonical'], subcell['consensus graph-based annotation (this study)']))\n",
    "\n",
    "# Standarization of the subcellular compartments to be merged with the compartments in the Sec Recon dataset\n",
    "for key in subcell_dict:\n",
    "    if subcell_dict[key] == 'early_endosome':\n",
    "        subcell_dict[key] = 'Early Endosome'\n",
    "    elif subcell_dict[key] == 'centrosome':\n",
    "        subcell_dict[key] = 'Centrosome'\n",
    "    elif subcell_dict[key] == 'ER':\n",
    "        subcell_dict[key] = 'Endoplasmic Reticulum'\n",
    "    elif subcell_dict[key] == 'mitochondrion':\n",
    "        subcell_dict[key] = 'Mitochondria'\n",
    "    elif subcell_dict[key] == 'stress_granule':\n",
    "        subcell_dict[key] = 'Stress Granule'\n",
    "    elif subcell_dict[key] == 'unclassified':\n",
    "        subcell_dict[key] = None\n",
    "    elif subcell_dict[key] == 'peroxisome':\n",
    "        subcell_dict[key] = 'Peroxisome'\n",
    "    elif subcell_dict[key] == '14-3-3_scaffold':\n",
    "        subcell_dict[key] = None\n",
    "    elif subcell_dict[key] == 'recycling_endosome':\n",
    "        subcell_dict[key] = 'Recycling Endosome'\n",
    "    elif subcell_dict[key] == 'plasma_membrane':\n",
    "        subcell_dict[key] = 'Plasma Membrane'\n",
    "    elif subcell_dict[key] == 'lysosome':\n",
    "        subcell_dict[key] = 'Lysosome'\n",
    "    elif subcell_dict[key] == 'translation':\n",
    "        subcell_dict[key] = 'Translation'\n",
    "    elif subcell_dict[key] == 'actin_cytoskeleton':\n",
    "        subcell_dict[key] = 'Actin Cytoskeleton'\n",
    "    elif subcell_dict[key] == 'cytosol':\n",
    "        subcell_dict[key] = 'Cytosol'\n",
    "    elif subcell_dict[key] == 'nucleus':\n",
    "        subcell_dict[key] = 'Nucleus'\n",
    "    elif subcell_dict[key] == 'ERGIC':\n",
    "        subcell_dict[key] = 'ERGIC'\n",
    "    elif subcell_dict[key] == 'p-body':\n",
    "        subcell_dict[key] = 'P-Body'\n",
    "    elif subcell_dict[key] == 'trans-Golgi':\n",
    "        subcell_dict[key] = 'trans-Golgi'\n",
    "    elif subcell_dict[key] == 'nucleolus':\n",
    "        subcell_dict[key] = 'Nucleolus'\n",
    "    elif subcell_dict[key] == 'proteasome':\n",
    "        subcell_dict[key] = 'Proteasome'\n",
    "    elif subcell_dict[key] == 'Golgi':\n",
    "        subcell_dict[key] = 'Golgi'\n",
    "\n",
    "# Map subcellular localization to the dataset\n",
    "sec_recon_dc['Subcellular Localization'] = sec_recon_dc['GENE SYMBOL'].map(subcell_dict)\n",
    "\n",
    "# Update the Google Sheet file\n",
    "if not sec_recon_dc.equals(sec_recon):\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print('Google Sheet updated on Subcellular Localization from \"subcellular_localization.csv\" dataset')\n",
    "else:\n",
    "    print('Subcellular Localizations from \"subcellular_localization.csv\" dataset are up-to-date')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subcellular localization of A3GALT2 is ['Golgi']\n",
      "Subcellular localization of A4GALT is ['Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of A4GNT is ['Golgi']\n",
      "Subcellular localization of ABL2 is ['Actin Cytoskeleton']\n",
      "Subcellular localization of ABO is ['Golgi', 'Secreted']\n",
      "Subcellular localization of ACAP1 is ['Recycling Endosome']\n",
      "Subcellular localization of ACAP3 is ['Recycling Endosome']\n",
      "Subcellular localization of AGAP1 is ['Cytoplasm']\n",
      "Subcellular localization of AGAP2 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of AGAP3 is ['Cytoplasm']\n",
      "Subcellular localization of AGR2 is ['Secreted']\n",
      "Subcellular localization of AGR3 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of AKAP10 is ['Mitochondria', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of ALG10B is ['Plasma Membrane']\n",
      "Subcellular localization of ALG1L2 is ['Plasma Membrane']\n",
      "Subcellular localization of AP1M2 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of AP3B2 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of AP3D1 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of AP3S2 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of ARAF is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of ARAP2 is ['Cytoplasm']\n",
      "Subcellular localization of ARAP3 is ['Cytoplasm']\n",
      "Subcellular localization of ARF3 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of ARFGAP2 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of ARHGAP24 is ['Actin Cytoskeleton']\n",
      "Subcellular localization of ARHGEF1 is ['Plasma Membrane']\n",
      "Subcellular localization of ARHGEF2 is ['Golgi', 'Actin Cytoskeleton', 'Cytoplasm']\n",
      "Subcellular localization of ARPC1B is ['Nucleus', 'Actin Cytoskeleton']\n",
      "Subcellular localization of ART1 is []\n",
      "Subcellular localization of ASAP3 is ['Cytoplasm']\n",
      "Subcellular localization of ATF4 is ['Nucleus', 'Actin Cytoskeleton', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of ATP2A1 is ['Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of ATP2A3 is ['Nucleus', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of ATXN3L is ['Nucleus']\n",
      "Subcellular localization of AXL is ['Plasma Membrane']\n",
      "Subcellular localization of B3GALNT1 is ['Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of B3GALNT2 is ['Golgi', 'Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of B3GALT1 is ['Golgi']\n",
      "Subcellular localization of B3GALT2 is ['Golgi']\n",
      "Subcellular localization of B3GALT4 is ['Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of B3GALT5 is ['Golgi']\n",
      "Subcellular localization of B3GAT1 is ['Golgi', 'Secreted']\n",
      "Subcellular localization of B3GAT2 is ['Golgi']\n",
      "Subcellular localization of B3GNT2 is ['Golgi']\n",
      "Subcellular localization of B3GNT3 is ['Golgi']\n",
      "Subcellular localization of B3GNT4 is ['Golgi']\n",
      "Subcellular localization of B3GNT5 is ['Golgi']\n",
      "Subcellular localization of B3GNT6 is ['Golgi']\n",
      "Subcellular localization of B3GNT7 is ['Golgi']\n",
      "Subcellular localization of B3GNT8 is ['Golgi']\n",
      "Google Sheet updated on Subcellular Localizations from Uniprot after 50 updates\n",
      "Subcellular localization of B3GNT9 is ['Golgi']\n",
      "Subcellular localization of B3GNTL1 is ['Golgi']\n",
      "Subcellular localization of B4GALNT1 is ['Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of B4GALNT2 is ['Golgi']\n",
      "Subcellular localization of B4GALNT3 is ['Golgi']\n",
      "Subcellular localization of B4GALNT4 is ['Golgi']\n",
      "Subcellular localization of B4GALT2 is ['Golgi']\n",
      "Subcellular localization of B4GALT4 is ['Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of B4GALT6 is ['Golgi']\n",
      "Subcellular localization of BAG2 is ['Golgi']\n",
      "Subcellular localization of BCL2 is ['Mitochondria', 'Cytoplasm', 'Endoplasmic Reticulum', 'Nucleus']\n",
      "Subcellular localization of CALM1 is ['Actin Cytoskeleton']\n",
      "Subcellular localization of CAPN2 is ['Plasma Membrane']\n",
      "Subcellular localization of CHST1 is ['Golgi']\n",
      "Subcellular localization of CHST10 is ['Golgi']\n",
      "Subcellular localization of CHST11 is ['Golgi']\n",
      "Subcellular localization of CHST13 is ['Golgi']\n",
      "Subcellular localization of CHST15 is ['Golgi']\n",
      "Subcellular localization of CHST2 is ['Plasma Membrane']\n",
      "Subcellular localization of CHST3 is ['Golgi']\n",
      "Subcellular localization of CHST4 is ['Golgi']\n",
      "Subcellular localization of CHST5 is ['Golgi']\n",
      "Subcellular localization of CHST6 is ['Plasma Membrane']\n",
      "Subcellular localization of CHST7 is ['Golgi']\n",
      "Subcellular localization of CHST8 is ['Golgi']\n",
      "Subcellular localization of CHST9 is ['Golgi']\n",
      "Subcellular localization of CHSY3 is ['Golgi']\n",
      "Subcellular localization of COLGALT2 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of COPZ2 is ['Endoplasmic Reticulum', 'Golgi', 'Cytoplasm']\n",
      "Subcellular localization of CRYAA is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of CRYAB is ['Lysosome', 'Nucleus', 'Cytoplasm', 'Secreted']\n",
      "Subcellular localization of CSGALNACT1 is ['Golgi']\n",
      "Subcellular localization of CSGALNACT2 is ['Golgi']\n",
      "Subcellular localization of CSNK1A1 is ['Nucleus', 'Actin Cytoskeleton', 'Cytoplasm']\n",
      "Subcellular localization of CSNK1A1L is ['Cytoplasm']\n",
      "Subcellular localization of CUL1 is ['Cytoplasm']\n",
      "Subcellular localization of CYTH2 is ['Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of CYTH3 is ['Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of CYTH4 is ['Plasma Membrane']\n",
      "Subcellular localization of DDI1 is ['Plasma Membrane']\n",
      "Subcellular localization of DDIT3 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of DERL3 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of DIPK1B is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of DIPK1C is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of DNAJA1 is ['Nucleus', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of DNAJA2 is ['Plasma Membrane']\n",
      "Subcellular localization of DNAJB2 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of DNAJB9 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of DNAJC5B is ['Plasma Membrane']\n",
      "Subcellular localization of DNAJC5G is ['Plasma Membrane']\n",
      "Google Sheet updated on Subcellular Localizations from Uniprot after 50 updates\n",
      "Subcellular localization of DNAJC6 is ['Plasma Membrane']\n",
      "Subcellular localization of DOLK is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of DPM2 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of DPP6 is ['Plasma Membrane']\n",
      "Subcellular localization of EDEM1 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of EDEM2 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of EIF2AK4 is ['Cytoplasm']\n",
      "Subcellular localization of ERP27 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of EXT1 is ['Golgi', 'cis-Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of EXTL1 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of FAM20A is ['Endoplasmic Reticulum', 'Golgi', 'Secreted']\n",
      "Subcellular localization of FAM20C is ['Endoplasmic Reticulum', 'Golgi', 'Secreted']\n",
      "Subcellular localization of FBXO2 is ['Cytoplasm']\n",
      "Subcellular localization of FBXO6 is ['Cytoplasm']\n",
      "Subcellular localization of FBXO8 is ['Cytoplasm']\n",
      "Subcellular localization of FKTN is ['Nucleus', 'Golgi', 'Cytoplasm']\n",
      "Subcellular localization of FOXO1 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of FUT1 is ['Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of FUT10 is ['Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of FUT2 is ['Golgi']\n",
      "Subcellular localization of FUT3 is ['Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of FUT4 is ['Golgi']\n",
      "Subcellular localization of FUT5 is ['Golgi']\n",
      "Subcellular localization of FUT6 is ['Golgi', 'Secreted']\n",
      "Subcellular localization of FUT7 is ['Golgi']\n",
      "Subcellular localization of FUT9 is ['trans-Golgi', 'Golgi']\n",
      "Subcellular localization of GAL3ST1 is ['Golgi']\n",
      "Subcellular localization of GAL3ST2 is ['Golgi']\n",
      "Subcellular localization of GAL3ST3 is ['Golgi']\n",
      "Subcellular localization of GAL3ST4 is ['Golgi']\n",
      "Subcellular localization of GALNT10 is ['Golgi']\n",
      "Subcellular localization of GALNT11 is ['Golgi']\n",
      "Subcellular localization of GALNT13 is ['Golgi']\n",
      "Subcellular localization of GALNT14 is ['Golgi']\n",
      "Subcellular localization of GALNT15 is ['Golgi']\n",
      "Subcellular localization of GALNT16 is ['Golgi']\n",
      "Subcellular localization of GALNT17 is ['Golgi']\n",
      "Subcellular localization of GALNT18 is ['Golgi']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subcellular localization of GALNT3 is ['Plasma Membrane']\n",
      "Subcellular localization of GALNT5 is ['Golgi']\n",
      "Subcellular localization of GALNT8 is ['Golgi']\n",
      "Subcellular localization of GALNT9 is ['Golgi']\n",
      "Subcellular localization of GALNT20 is ['Late Endosome']\n",
      "Subcellular localization of GANAB is ['Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of GBF1 is ['Golgi']\n",
      "Subcellular localization of GBGT1 is ['Golgi']\n",
      "Subcellular localization of GCNT1 is ['Golgi']\n",
      "Subcellular localization of GCNT2 is ['Golgi']\n",
      "Subcellular localization of GCNT3 is ['Golgi']\n",
      "Subcellular localization of GCNT4 is ['Golgi']\n",
      "Google Sheet updated on Subcellular Localizations from Uniprot after 50 updates\n",
      "Subcellular localization of GCNT7 is ['Golgi']\n",
      "Subcellular localization of GGA3 is ['trans-Golgi', 'Early Endosome', 'Plasma Membrane', 'Endosome']\n",
      "Subcellular localization of GLT1D1 is ['Secreted']\n",
      "Subcellular localization of GLT6D1 is ['Plasma Membrane']\n",
      "Subcellular localization of GLT8D2 is ['Plasma Membrane']\n",
      "Subcellular localization of GNPTAB is ['Golgi']\n",
      "Subcellular localization of GOSR1 is ['Golgi']\n",
      "Subcellular localization of GPLD1 is ['Secreted']\n",
      "Subcellular localization of GXYLT2 is ['Plasma Membrane']\n",
      "Subcellular localization of HAS1 is ['Plasma Membrane']\n",
      "Subcellular localization of HAS2 is ['Lysosome', 'Golgi', 'Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of HAS3 is ['trans-Golgi', 'Early Endosome', 'Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of HS3ST1 is ['Golgi']\n",
      "Subcellular localization of HS3ST2 is ['Golgi']\n",
      "Subcellular localization of HS3ST3A1 is ['Golgi']\n",
      "Subcellular localization of HS3ST3B1 is ['Golgi']\n",
      "Subcellular localization of HS3ST4 is ['Golgi']\n",
      "Subcellular localization of HS3ST5 is ['Golgi']\n",
      "Subcellular localization of HS3ST6 is ['Golgi']\n",
      "Subcellular localization of HS6ST1 is ['Plasma Membrane']\n",
      "Subcellular localization of HS6ST3 is ['Plasma Membrane']\n",
      "Subcellular localization of HSPA12B is ['Plasma Membrane']\n",
      "Subcellular localization of HSPA1B is ['Actin Cytoskeleton']\n",
      "Subcellular localization of HSPA1L is ['Actin Cytoskeleton']\n",
      "Subcellular localization of HSPA8 is ['Lysosome', 'Cytoplasm', 'Nucleolus', 'Plasma Membrane']\n",
      "Subcellular localization of HSPH1 is ['Cytoplasm']\n",
      "Subcellular localization of HYOU1 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of INPP5E is ['Nucleus', 'Golgi', 'Actin Cytoskeleton', 'Plasma Membrane', 'Cytoplasm']\n",
      "Subcellular localization of IQSEC1 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of IQSEC2 is ['Cytoplasm']\n",
      "Subcellular localization of IQSEC3 is ['Cytoplasm']\n",
      "Subcellular localization of KCNN2 is ['Plasma Membrane']\n",
      "Subcellular localization of KDELR2 is ['Golgi', 'Endoplasmic Reticulum', 'Cytoplasm']\n",
      "Subcellular localization of KDELR3 is ['Golgi', 'Endoplasmic Reticulum', 'Cytoplasm']\n",
      "Subcellular localization of KIF23 is ['Nucleus', 'Actin Cytoskeleton']\n",
      "Subcellular localization of KNCN is ['Plasma Membrane']\n",
      "Subcellular localization of LARGE1 is ['Plasma Membrane']\n",
      "Subcellular localization of LARGE2 is ['Golgi']\n",
      "Subcellular localization of LFNG is ['Golgi']\n",
      "Subcellular localization of LGALS9 is ['Nucleus', 'Cytoplasm', 'Secreted']\n",
      "Subcellular localization of LMAN1L is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of MAN1C1 is ['Golgi']\n",
      "Subcellular localization of MAN2A2 is ['Golgi']\n",
      "Subcellular localization of MAP2K7 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of MAP3K5 is ['Endoplasmic Reticulum', 'Cytoplasm']\n",
      "Subcellular localization of MAPK10 is ['Nucleus', 'Cytoplasm', 'Mitochondria', 'Plasma Membrane']\n",
      "Subcellular localization of MFNG is ['Golgi']\n",
      "Subcellular localization of MGAT3 is ['Golgi']\n",
      "Subcellular localization of MGAT4A is ['Golgi']\n",
      "Subcellular localization of MGAT4C is ['Golgi']\n",
      "Google Sheet updated on Subcellular Localizations from Uniprot after 50 updates\n",
      "Subcellular localization of MGAT5B is ['Golgi']\n",
      "Subcellular localization of MICALL2 is ['Recycling Endosome', 'Actin Cytoskeleton', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of MPPE1 is ['cis-Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of MYO5B is ['Cytoplasm']\n",
      "Subcellular localization of NAGPA is ['trans-Golgi', 'Golgi']\n",
      "Subcellular localization of NAPB is ['Plasma Membrane']\n",
      "Subcellular localization of NAT8 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of NDE1 is ['Actin Cytoskeleton']\n",
      "Subcellular localization of NDST2 is ['Golgi']\n",
      "Subcellular localization of NDST3 is ['Golgi']\n",
      "Subcellular localization of NDST4 is ['Golgi']\n",
      "Subcellular localization of NET1 is ['Cytoplasm']\n",
      "Subcellular localization of NEU2 is ['Cytoplasm']\n",
      "Subcellular localization of NEU4 is ['Mitochondria', 'Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of NSFL1C is ['Nucleus', 'Golgi', 'Actin Cytoskeleton']\n",
      "Subcellular localization of OGT is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of PDIA2 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of PDIA3 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of PDIA4 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of PDILT is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of PGAP2 is ['Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of PGAP3 is ['Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of PGAP4 is ['Golgi']\n",
      "Subcellular localization of PIGB is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of PIGC is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of PIGF is ['Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of PIGL is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of PIGP is ['Plasma Membrane']\n",
      "Subcellular localization of PIGQ is ['Plasma Membrane']\n",
      "Subcellular localization of PIGV is ['Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of PIGY is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of PIGZ is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of PIK3CA is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of PKD1 is ['Golgi', 'Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of PKDCC is ['Golgi', 'Secreted']\n",
      "Subcellular localization of PLAA is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of POMT1 is ['Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of PRKCSH is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of PRKN is ['Endoplasmic Reticulum', 'Nucleus', 'Cytoplasm', 'Mitochondria']\n",
      "Subcellular localization of PSD is ['Plasma Membrane']\n",
      "Subcellular localization of PSD2 is ['Plasma Membrane']\n",
      "Subcellular localization of PSD4 is ['Plasma Membrane']\n",
      "Subcellular localization of PSMA8 is ['Nucleus']\n",
      "Subcellular localization of PSMB10 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of PSMB11 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of PSMB8 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of PSMB9 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of PSMC3 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of PSMD4 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of PSMD5 is ['Nucleus', 'Cytoplasm']\n",
      "Google Sheet updated on Subcellular Localizations from Uniprot after 50 updates\n",
      "Subcellular localization of QSOX1 is ['Golgi', 'Secreted']\n",
      "Subcellular localization of RAB11A is ['trans-Golgi', 'Golgi', 'Endosome', 'Plasma Membrane', 'Cytoplasm', 'Recycling Endosome']\n",
      "Subcellular localization of RAB11FIP3 is ['trans-Golgi', 'Golgi', 'Actin Cytoskeleton', 'Endosome', 'Recycling Endosome']\n",
      "Subcellular localization of RAB11FIP4 is ['Recycling Endosome', 'Actin Cytoskeleton', 'Cytoplasm']\n",
      "Subcellular localization of RAB15 is ['Plasma Membrane']\n",
      "Subcellular localization of RAB17 is ['Recycling Endosome']\n",
      "Subcellular localization of RAB19 is ['Plasma Membrane']\n",
      "Subcellular localization of RAB20 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of RAB25 is ['Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of RAB26 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of RAB27A is ['Lysosome', 'Late Endosome']\n",
      "Subcellular localization of RAB27B is ['Late Endosome', 'Plasma Membrane']\n",
      "Subcellular localization of RAB33A is ['Plasma Membrane']\n",
      "Subcellular localization of RAB34 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of RAB36 is ['Golgi']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subcellular localization of RAB37 is ['Cytoplasm']\n",
      "Subcellular localization of RAB38 is ['Plasma Membrane']\n",
      "Subcellular localization of RAB3C is ['Plasma Membrane']\n",
      "Subcellular localization of RAB40A is ['Plasma Membrane']\n",
      "Subcellular localization of RAB40AL is ['Mitochondria', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of RAB40B is ['Plasma Membrane']\n",
      "Subcellular localization of RAB40C is ['Plasma Membrane']\n",
      "Subcellular localization of RAB41 is ['Cytoplasm']\n",
      "Subcellular localization of RAB42 is ['Plasma Membrane']\n",
      "Subcellular localization of RAB44 is ['Plasma Membrane']\n",
      "Subcellular localization of RAB6A is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of RAB6C is ['Nucleus', 'Actin Cytoskeleton', 'Cytoplasm']\n",
      "Subcellular localization of RABGAP1L is ['Early Endosome', 'Golgi', 'Cytoplasm']\n",
      "Subcellular localization of RASEF is ['Cytoplasm']\n",
      "Subcellular localization of RBX1 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of RNF185 is ['Mitochondria', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of RNF19A is ['Actin Cytoskeleton', 'Plasma Membrane']\n",
      "Subcellular localization of RRBP1 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of RXYLT1 is ['Golgi']\n",
      "Subcellular localization of RYR1 is []\n",
      "Subcellular localization of RYR2 is []\n",
      "Subcellular localization of RYR3 is []\n",
      "Subcellular localization of SEC16A is ['Golgi', 'Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of SEC16B is ['Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of SEC22B is ['trans-Golgi', 'cis-Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of SEC31B is ['Endoplasmic Reticulum', 'ERGIC', 'Cytoplasm']\n",
      "Subcellular localization of SEM1 is ['Nucleus']\n",
      "Subcellular localization of SERP1 is ['Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of SGSM1 is ['trans-Golgi', 'Cytoplasm']\n",
      "Subcellular localization of SGSM2 is ['Cytoplasm']\n",
      "Subcellular localization of SGSM3 is ['Cytoplasm']\n",
      "Subcellular localization of SIAH1 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of SIAH2 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of SLC35A3 is ['Golgi']\n",
      "Subcellular localization of SLC35B3 is ['Golgi']\n",
      "Google Sheet updated on Subcellular Localizations from Uniprot after 50 updates\n",
      "Subcellular localization of SLC35B4 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of SLC35C1 is ['Golgi']\n",
      "Subcellular localization of SLC35C2 is ['Plasma Membrane']\n",
      "Subcellular localization of SLC35D1 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of SLC35D2 is ['Golgi']\n",
      "Subcellular localization of SNAP25 is ['Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of SRP14 is ['Cytoplasm']\n",
      "Subcellular localization of SRP68 is ['Endoplasmic Reticulum', 'Cytoplasm', 'Nucleolus']\n",
      "Subcellular localization of SRP72 is ['Endoplasmic Reticulum', 'Cytoplasm']\n",
      "Subcellular localization of SRP9 is ['Cytoplasm']\n",
      "Subcellular localization of SSR2 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of ST3GAL1 is ['trans-Golgi', 'Golgi', 'Secreted']\n",
      "Subcellular localization of ST3GAL2 is ['Golgi', 'Secreted']\n",
      "Subcellular localization of ST3GAL3 is ['Golgi', 'Secreted']\n",
      "Subcellular localization of ST3GAL4 is ['Plasma Membrane']\n",
      "Subcellular localization of ST3GAL5 is ['Golgi']\n",
      "Subcellular localization of ST3GAL6 is ['Golgi']\n",
      "Subcellular localization of ST6GAL1 is ['Golgi', 'Secreted']\n",
      "Subcellular localization of ST6GAL2 is ['Golgi']\n",
      "Subcellular localization of ST6GALNAC1 is ['Golgi']\n",
      "Subcellular localization of ST6GALNAC2 is ['Golgi']\n",
      "Subcellular localization of ST6GALNAC3 is ['Plasma Membrane']\n",
      "Subcellular localization of ST6GALNAC4 is ['Plasma Membrane']\n",
      "Subcellular localization of ST6GALNAC5 is ['Golgi']\n",
      "Subcellular localization of ST6GALNAC6 is ['Plasma Membrane']\n",
      "Subcellular localization of ST8SIA1 is ['Golgi']\n",
      "Subcellular localization of ST8SIA2 is ['Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of ST8SIA3 is ['Golgi']\n",
      "Subcellular localization of ST8SIA4 is ['Golgi']\n",
      "Subcellular localization of ST8SIA5 is ['Golgi']\n",
      "Subcellular localization of ST8SIA6 is ['Golgi']\n",
      "Subcellular localization of STX11 is ['trans-Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of STX16 is ['Golgi']\n",
      "Subcellular localization of STX19 is ['Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of STX1A is ['Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of STX1B is ['Plasma Membrane']\n",
      "Subcellular localization of SYTL2 is ['Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of TBC1D1 is ['Nucleus']\n",
      "Subcellular localization of TBC1D10C is ['Nucleus']\n",
      "Subcellular localization of TBC1D12 is []\n",
      "Subcellular localization of TBC1D14 is ['trans-Golgi', 'cis-Golgi']\n",
      "Subcellular localization of TBC1D15 is ['Cytoplasm']\n",
      "Subcellular localization of TBC1D2 is ['Cytoplasm']\n",
      "Subcellular localization of TBC1D22B is ['Cytoplasm']\n",
      "Subcellular localization of TBC1D30 is ['Plasma Membrane']\n",
      "Subcellular localization of TBC1D4 is ['Cytoplasm']\n",
      "Subcellular localization of TBC1D7 is ['Lysosome', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of TBC1D9 is ['Lysosome', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of TBC1D9B is ['Plasma Membrane']\n",
      "Subcellular localization of TMED6 is ['Endoplasmic Reticulum']\n",
      "Google Sheet updated on Subcellular Localizations from Uniprot after 50 updates\n",
      "Subcellular localization of TNK2 is ['Nucleus', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of TP53 is ['Nucleus', 'Mitochondria', 'Actin Cytoskeleton', 'Endoplasmic Reticulum', 'Cytoplasm']\n",
      "Subcellular localization of TPST1 is ['Golgi']\n",
      "Subcellular localization of TPST2 is ['Golgi']\n",
      "Subcellular localization of TRAP1 is ['Mitochondria']\n",
      "Subcellular localization of TUSC3 is ['Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of TXNDC5 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of UBA5 is ['Endoplasmic Reticulum', 'Nucleus', 'Golgi', 'Cytoplasm']\n",
      "Subcellular localization of UBA7 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of UBB is ['Nucleus', 'Cytoplasm', 'Mitochondria']\n",
      "Subcellular localization of UBE2D2 is ['Nucleus', 'Cytoplasm', 'Mitochondria']\n",
      "Subcellular localization of UBE2D4 is ['Nucleus', 'Cytoplasm', 'Mitochondria']\n",
      "Subcellular localization of UBE2E3 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of UBE2Z is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of UBQLN1 is ['Endoplasmic Reticulum', 'Nucleus', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of UBQLN3 is ['Endoplasmic Reticulum', 'Nucleus', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of UBQLNL is ['Endoplasmic Reticulum', 'Nucleus', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of UBXN2A is ['Nucleus', 'Golgi', 'Endoplasmic Reticulum', 'Cytoplasm']\n",
      "Subcellular localization of UFD1 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of UGCG is ['Golgi']\n",
      "Subcellular localization of UGT3A1 is ['Plasma Membrane']\n",
      "Subcellular localization of USP6 is ['Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of UST is ['Golgi']\n",
      "Subcellular localization of VAMP1 is ['Cytoplasm']\n",
      "Subcellular localization of VAMP5 is ['Plasma Membrane']\n",
      "Subcellular localization of VCP is ['Nucleus', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of WAS is ['Nucleus']\n",
      "Subcellular localization of WDR38 is ['Nucleus']\n",
      "Subcellular localization of XBP1 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of XYLT1 is ['Golgi', 'Secreted']\n",
      "Subcellular localization of USP19 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of DGAT2 is ['Cytoplasm', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of TRAPPC3L is ['cis-Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of SEC22C is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of BAP29 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of BECLIN-1 is ['Mitochondria', 'trans-Golgi', 'Endoplasmic Reticulum', 'Endosome', 'Cytoplasm']\n",
      "Subcellular localization of BIM is []\n",
      "Subcellular localization of BTK is ['Cytoplasm']\n",
      "Subcellular localization of CAPN10 is ['Cytoplasm']\n",
      "Subcellular localization of CAPN11 is ['Cytoplasm']\n",
      "Subcellular localization of CAPN12 is ['Cytoplasm']\n",
      "Subcellular localization of CAPN13 is ['Cytoplasm']\n",
      "Subcellular localization of CAPN3 is ['Cytoplasm', 'Nucleolus']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subcellular localization of CAPN6 is ['Actin Cytoskeleton', 'Cytoplasm']\n",
      "Subcellular localization of CAPN8 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of CAPN9 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of CASP12 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of CASQ1 is ['Mitochondria', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of CASQ2 is []\n",
      "Subcellular localization of CLPB is ['Mitochondria']\n",
      "Google Sheet updated on Subcellular Localizations from Uniprot after 50 updates\n",
      "Subcellular localization of CNIH is ['Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of CNIH2 is ['Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of CNIH3 is []\n",
      "Subcellular localization of COG3 is ['Golgi']\n",
      "Subcellular localization of COG5 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of COPS1 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of CALR3 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of DNAJA4 is ['Plasma Membrane']\n",
      "Subcellular localization of DNAJB5 is ['Plasma Membrane']\n",
      "Subcellular localization of DNAJB6 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of DNAJB7 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of DNAJB8 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of DNAJC8 is ['Nucleus']\n",
      "Subcellular localization of DR5 is ['Plasma Membrane']\n",
      "Subcellular localization of ERFAD is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of EXOC3L1 is ['Cytoplasm']\n",
      "Subcellular localization of EXOC3L4 is ['Cytoplasm']\n",
      "Subcellular localization of FBXO1 is ['Nucleus', 'Actin Cytoskeleton', 'Cytoplasm']\n",
      "Subcellular localization of FKBP11 is ['Plasma Membrane']\n",
      "Subcellular localization of FKBP6 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of FKBP8 is ['Mitochondria']\n",
      "Subcellular localization of FMO1 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of FRAP is ['Lysosome', 'Mitochondria', 'Nucleus', 'Golgi', 'Endoplasmic Reticulum', 'Cytoplasm']\n",
      "Subcellular localization of FURIN is ['trans-Golgi', 'Endosome', 'Secreted', 'Plasma Membrane']\n",
      "Subcellular localization of HERP2 is ['Nucleus']\n",
      "Subcellular localization of HSPB8 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of IRE1B is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of IRS1 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of JKAMP1 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of MPR is ['Mitochondria', 'Secreted']\n",
      "Subcellular localization of NEDD4 is ['Nucleus', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of NEDD4L is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of OST4 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of OSTC is ['Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of OXA1L is ['Mitochondria']\n",
      "Subcellular localization of PACE1 is ['Golgi', 'Cytoplasm']\n",
      "Subcellular localization of PACS-2 is ['Mitochondria', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of PPIL2 is ['Nucleus']\n",
      "Subcellular localization of PPIL4 is ['Nucleus']\n",
      "Subcellular localization of PPIL6 is ['Nucleus']\n",
      "Subcellular localization of PSME3 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of RACK1 is ['Nucleus', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of RNF103 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of RNF77 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of S1P is ['Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of S2P is ['Plasma Membrane']\n",
      "Subcellular localization of SAKS1 is ['Cytoplasm']\n",
      "Subcellular localization of SEH1 is ['Lysosome', 'Nucleus']\n",
      "Subcellular localization of SELK is ['Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of Sep15 is ['Endoplasmic Reticulum']\n",
      "Google Sheet updated on Subcellular Localizations from Uniprot after 50 updates\n",
      "Subcellular localization of SNF8 is ['Nucleus', 'Cytoplasm', 'Late Endosome', 'Endosome']\n",
      "Subcellular localization of SORCS1 is ['Plasma Membrane']\n",
      "Subcellular localization of SORCS2 is ['Early Endosome', 'Recycling Endosome', 'Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of SORCS3 is ['Plasma Membrane']\n",
      "Subcellular localization of STXBP6 is ['Cytoplasm', 'Plasma Membrane']\n",
      "Subcellular localization of SYS1 is ['Golgi']\n",
      "Subcellular localization of TAPBP is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of CCT6B is ['Cytoplasm']\n",
      "Subcellular localization of TRAM1L1 is ['Endoplasmic Reticulum']\n",
      "Subcellular localization of TRAM2 is ['Plasma Membrane']\n",
      "Subcellular localization of TRIM21 is ['Nucleus', 'Cytoplasm']\n",
      "Subcellular localization of TRPV4 is ['Plasma Membrane']\n",
      "Subcellular localization of UBE2D2 is ['Plasma Membrane']\n",
      "Subcellular localization of UBR5 is ['Nucleus']\n",
      "Subcellular localization of UBXN7 is ['Nucleus']\n",
      "Subcellular localization of VAMPA is ['Nucleus', 'Endoplasmic Reticulum', 'Plasma Membrane']\n",
      "Subcellular localization of VCIP135 is ['Endoplasmic Reticulum', 'Nucleus', 'Golgi', 'Cytoplasm']\n",
      "Subcellular localization of VIPL is ['Golgi', 'Endoplasmic Reticulum']\n",
      "Subcellular localization of VPS36 is ['Late Endosome', 'Cytoplasm', 'Nucleus', 'Plasma Membrane']\n",
      "Subcellular localization of VPS37D is ['Late Endosome']\n",
      "Subcellular localization of Vps52 is ['trans-Golgi', 'Recycling Endosome', 'Endosome']\n",
      "Subcellular localization of Vps53 is ['trans-Golgi', 'Recycling Endosome', 'Endosome']\n",
      "Subcellular localization of Vps54 is ['trans-Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of VPS9D1 is ['trans-Golgi', 'Plasma Membrane']\n",
      "Subcellular localization of VPS51 is ['trans-Golgi', 'Recycling Endosome']\n",
      "Google Sheet updated on Subcellular Localizations from Uniprot after 25 updates\n"
     ]
    }
   ],
   "source": [
    "#Retrieval of Subcellular localizations from Uniprot\n",
    "\n",
    "loop_counter = 0\n",
    "update_threshold = 50\n",
    "\n",
    "for i, row in sec_recon_dc.iterrows():\n",
    "    gene = row['GENE SYMBOL']\n",
    "    # Subcellular compartments are extracted using the Human Uniprot ID\n",
    "    uniprot_ids = row['HUMAN UNIPROT'].split(\", \")\n",
    "    if (pd.isna(row['Subcellular Localization']) or row['Subcellular Localization'] == ''):\n",
    "         if uniprot_ids != ['']:\n",
    "            for uni_id in uniprot_ids:\n",
    "                sub_loc = Request_Utilis.get_subcellular_localization(uni_id)\n",
    "                if sub_loc is not None:\n",
    "                    new_sub_loc = []\n",
    "                    for sloc in sub_loc:\n",
    "                        # Standarization of the subcellular compartments to be included in the Sec Recon dataset\n",
    "                        match_found = False\n",
    "                        if sloc.startswith('Recycling endosome'):\n",
    "                            sloc = 'Recycling Endosome'\n",
    "                            match_found = True\n",
    "                        if sloc.startswith('Late endosome'):\n",
    "                            sloc = 'Late Endosome'\n",
    "                            match_found = True\n",
    "                        if sloc.startswith('Endosome membrane'):\n",
    "                            sloc = 'Endosome'\n",
    "                            match_found = True\n",
    "                        if sloc.startswith('Early endosome'):\n",
    "                            sloc = 'Early Endosome'\n",
    "                            match_found = True\n",
    "                        elif sloc.startswith('Endoplasmic Reticulum-Golgi'): \n",
    "                            sloc = 'ERGIC'    \n",
    "                            match_found = True\n",
    "                        elif sloc.startswith('Endoplasmic reticulum'):\n",
    "                            sloc = 'Endoplasmic Reticulum'\n",
    "                            match_found = True\n",
    "                        elif 'COPII' in sloc:\n",
    "                            sloc = 'ERGIC'    \n",
    "                            match_found = True\n",
    "                        elif 'cytoskeleton' in sloc:\n",
    "                            sloc = 'Actin Cytoskeleton'\n",
    "                            match_found = True\n",
    "                        elif sloc.startswith('Cytoplasm'):\n",
    "                            sloc = 'Cytoplasm'\n",
    "                            match_found = True\n",
    "                        elif 'trans-Golgi' in sloc:\n",
    "                            sloc = 'trans-Golgi'\n",
    "                            match_found = True\n",
    "                        elif 'cis-Golgi' in sloc:\n",
    "                            sloc = 'cis-Golgi'\n",
    "                            match_found = True\n",
    "                        elif sloc.startswith('Golgi apparatus'):\n",
    "                            sloc = 'Golgi'\n",
    "                            match_found = True\n",
    "                        elif 'nucleolus' in sloc:\n",
    "                            sloc = 'Nucleolus'\n",
    "                            match_found = True\n",
    "                        elif sloc.startswith('Nucleus'):\n",
    "                            sloc = 'Nucleus'\n",
    "                            match_found = True\n",
    "                        elif sloc.startswith('Mitochondrion'):\n",
    "                            sloc = 'Mitochondria'\n",
    "                            match_found = True\n",
    "                        elif sloc == 'Membrane' or sloc == 'Cell membrane':\n",
    "                            sloc = 'Plasma Membrane'\n",
    "                            match_found = True\n",
    "                        elif sloc.startswith('Lysosome'):\n",
    "                            sloc = 'Lysosome'\n",
    "                            match_found = True\n",
    "                        elif sloc == 'Secreted':\n",
    "                            match_found = True\n",
    "                        if not match_found:\n",
    "                            continue\n",
    "                            \n",
    "                        new_sub_loc.append(sloc)\n",
    "                            \n",
    "                    break\n",
    "            print(f'Subcellular localization of {gene} is {list(set(new_sub_loc))}')\n",
    "            sec_recon_dc.at[i, 'Subcellular Localization'] = list(set(new_sub_loc))\n",
    "            loop_counter += 1\n",
    "            \n",
    "            # After 50 iterations of the loop, update the Google Sheet file\n",
    "            if loop_counter >= update_threshold:\n",
    "                if not sec_recon_dc.equals(sec_recon):\n",
    "                    sec_recon_dc['Subcellular Localization'] = sec_recon_dc['Subcellular Localization'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)\n",
    "                    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "                    print(f\"Google Sheet updated on Subcellular Localizations from Uniprot after {loop_counter} updates\")\n",
    "                else:\n",
    "                    print('Subcellular Localizations from Uniprot are up-to-date')\n",
    "                loop_counter = 0\n",
    "\n",
    "# Check if there are any remaining updates after exiting the loop\n",
    "if loop_counter > 0 and not sec_recon_dc.equals(sec_recon):\n",
    "    sec_recon_dc['Subcellular Localization'] = sec_recon_dc['Subcellular Localization'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(f\"Google Sheet updated on Subcellular Localizations from Uniprot after {loop_counter} updates\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
