{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get Gene Info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "import Request_Utilis\n",
    "from google_sheet import GoogleSheet\n",
    "\n",
    "Entrez.email = \"a.antonakoudis@sartorius.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE SYMBOL</th>\n",
       "      <th>ALIAS</th>\n",
       "      <th>ENSEMBL</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>GENENAME</th>\n",
       "      <th>Mapped Process</th>\n",
       "      <th>SecM</th>\n",
       "      <th>Regulator</th>\n",
       "      <th>SecP</th>\n",
       "      <th>Membrane</th>\n",
       "      <th>...</th>\n",
       "      <th>Process 8 Ref</th>\n",
       "      <th>Process 8 Genes</th>\n",
       "      <th>Process 9</th>\n",
       "      <th>Process 9 Mean Score</th>\n",
       "      <th>Process 9 Score SD</th>\n",
       "      <th>Process 9 Score Count</th>\n",
       "      <th>Process 9 Raw Scores</th>\n",
       "      <th>Process 9 Curators</th>\n",
       "      <th>Process 9 Ref</th>\n",
       "      <th>Process 9 Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>A3GALT2P,IGB3S,IGBS3S,A3GALT2</td>\n",
       "      <td>ENSG00000184389</td>\n",
       "      <td>127550</td>\n",
       "      <td>alpha 1,3-galactosyltransferase 2</td>\n",
       "      <td>Glycosylation', 'Glycosylation'</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>A14GALT,A4GALT1,Gb3S,P(k),P1,P1PK,PK,A4GALT</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>53947</td>\n",
       "      <td>alpha 1,4-galactosyltransferase (P blood group)</td>\n",
       "      <td>glycolipid, globo series', 'globo series', 'gl...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A4GNT</td>\n",
       "      <td>alpha4GnT,A4GNT</td>\n",
       "      <td>ENSG00000118017</td>\n",
       "      <td>51146</td>\n",
       "      <td>alpha-1,4-N-acetylglucosaminyltransferase</td>\n",
       "      <td>O-glycan', 'O-glycan, mucin-type', 'mucin-type'</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>ABL,BCR-ABL,CHDSKM,JTK7,bcr/abl,c-ABL,c-ABL1,p...</td>\n",
       "      <td>ENSG00000097007</td>\n",
       "      <td>25</td>\n",
       "      <td>ABL proto-oncogene 1, non-receptor tyrosine ki...</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABL2</td>\n",
       "      <td>ABLL,ARG,ABL2</td>\n",
       "      <td>ENSG00000143322</td>\n",
       "      <td>27</td>\n",
       "      <td>ABL proto-oncogene 2, non-receptor tyrosine ki...</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>911</th>\n",
       "      <td>YKT6</td>\n",
       "      <td>YKT6</td>\n",
       "      <td>ENSG00000106636</td>\n",
       "      <td>10652</td>\n",
       "      <td>YKT6 v-SNARE homolog</td>\n",
       "      <td>General Trafficking', 'General Trafficking', '...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>912</th>\n",
       "      <td>YOD1</td>\n",
       "      <td>DUBA8,OTUD2,PRO0907,YOD1</td>\n",
       "      <td>ENSG00000180667</td>\n",
       "      <td>55432</td>\n",
       "      <td>YOD1 deubiquitinase</td>\n",
       "      <td>ubl conjugation pathway', 'ERAD', 'ERAD', 'ERA...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>913</th>\n",
       "      <td>USP19</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10869</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914</th>\n",
       "      <td>DGAT1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8694</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>915</th>\n",
       "      <td>DGAT2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>84649</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>916 rows × 86 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    GENE SYMBOL                                              ALIAS  \\\n",
       "0       A3GALT2                      A3GALT2P,IGB3S,IGBS3S,A3GALT2   \n",
       "1        A4GALT        A14GALT,A4GALT1,Gb3S,P(k),P1,P1PK,PK,A4GALT   \n",
       "2         A4GNT                                    alpha4GnT,A4GNT   \n",
       "3          ABL1  ABL,BCR-ABL,CHDSKM,JTK7,bcr/abl,c-ABL,c-ABL1,p...   \n",
       "4          ABL2                                      ABLL,ARG,ABL2   \n",
       "..          ...                                                ...   \n",
       "911        YKT6                                               YKT6   \n",
       "912        YOD1                           DUBA8,OTUD2,PRO0907,YOD1   \n",
       "913       USP19                                                      \n",
       "914       DGAT1                                                      \n",
       "915       DGAT2                                                      \n",
       "\n",
       "             ENSEMBL ENTREZID  \\\n",
       "0    ENSG00000184389   127550   \n",
       "1    ENSG00000128274    53947   \n",
       "2    ENSG00000118017    51146   \n",
       "3    ENSG00000097007       25   \n",
       "4    ENSG00000143322       27   \n",
       "..               ...      ...   \n",
       "911  ENSG00000106636    10652   \n",
       "912  ENSG00000180667    55432   \n",
       "913                     10869   \n",
       "914                      8694   \n",
       "915                     84649   \n",
       "\n",
       "                                              GENENAME  \\\n",
       "0                    alpha 1,3-galactosyltransferase 2   \n",
       "1      alpha 1,4-galactosyltransferase (P blood group)   \n",
       "2            alpha-1,4-N-acetylglucosaminyltransferase   \n",
       "3    ABL proto-oncogene 1, non-receptor tyrosine ki...   \n",
       "4    ABL proto-oncogene 2, non-receptor tyrosine ki...   \n",
       "..                                                 ...   \n",
       "911                               YKT6 v-SNARE homolog   \n",
       "912                                YOD1 deubiquitinase   \n",
       "913                                                      \n",
       "914                                                      \n",
       "915                                                      \n",
       "\n",
       "                                        Mapped Process SecM Regulator SecP  \\\n",
       "0                      Glycosylation', 'Glycosylation'    0         0    0   \n",
       "1    glycolipid, globo series', 'globo series', 'gl...    0         0    0   \n",
       "2      O-glycan', 'O-glycan, mucin-type', 'mucin-type'    0         0    0   \n",
       "3                                                   []    1         1    0   \n",
       "4                                                   []    1         1    0   \n",
       "..                                                 ...  ...       ...  ...   \n",
       "911  General Trafficking', 'General Trafficking', '...    0         0    0   \n",
       "912  ubl conjugation pathway', 'ERAD', 'ERAD', 'ERA...    2         1    0   \n",
       "913                                                       1         1    0   \n",
       "914                                                       1         1    0   \n",
       "915                                                       1         1    0   \n",
       "\n",
       "    Membrane  ... Process 8 Ref Process 8 Genes Process 9  \\\n",
       "0          0  ...          None            None      None   \n",
       "1          0  ...          None            None      None   \n",
       "2          0  ...          None            None      None   \n",
       "3          0  ...          None            None      None   \n",
       "4          0  ...          None            None      None   \n",
       "..       ...  ...           ...             ...       ...   \n",
       "911        1  ...          None            None      None   \n",
       "912        0  ...          None            None      None   \n",
       "913        0  ...          None            None      None   \n",
       "914        1  ...          None            None      None   \n",
       "915        1  ...          None            None      None   \n",
       "\n",
       "    Process 9 Mean Score Process 9 Score SD Process 9 Score Count  \\\n",
       "0                   None               None                  None   \n",
       "1                   None               None                  None   \n",
       "2                   None               None                  None   \n",
       "3                   None               None                  None   \n",
       "4                   None               None                  None   \n",
       "..                   ...                ...                   ...   \n",
       "911                 None               None                  None   \n",
       "912                 None               None                  None   \n",
       "913                 None               None                  None   \n",
       "914                 None               None                  None   \n",
       "915                 None               None                  None   \n",
       "\n",
       "    Process 9 Raw Scores Process 9 Curators Process 9 Ref Process 9 Genes  \n",
       "0                   None               None          None            None  \n",
       "1                   None               None          None            None  \n",
       "2                   None               None          None            None  \n",
       "3                   None               None          None            None  \n",
       "4                   None               None          None            None  \n",
       "..                   ...                ...           ...             ...  \n",
       "911                 None               None          None            None  \n",
       "912                 None               None          None            None  \n",
       "913                 None               None          None            None  \n",
       "914                 None               None          None            None  \n",
       "915                 None               None          None            None  \n",
       "\n",
       "[916 rows x 86 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from google_sheet import GoogleSheet\n",
    "\n",
    "##### ----- Generate datasets from Google Sheet ----- #####\n",
    "\n",
    "#Credential file\n",
    "KEY_FILE_PATH = 'credentials.json'\n",
    "\n",
    "#CHO Network Reconstruction + Recon3D_v3 Google Sheet ID\n",
    "SPREADSHEET_ID = '1DaAdZlvMYDqb7g31I5dw-ZCZH52Xj_W3FnQMFUzqmiQ'\n",
    "\n",
    "# Initialize the GoogleSheet object\n",
    "gsheet_file = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "\n",
    "# Read data from the Google Sheet\n",
    "sec_recon_sheet = 'Sheet1'\n",
    "sec_recon = gsheet_file.read_google_sheet(sec_recon_sheet)\n",
    "sec_recon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sec_recon_dc = sec_recon.copy()\n",
    "for i,row in sec_recon_dc.iterrows():\n",
    "    if pd.isnull(row['ENTREZID']) or row['ENTREZID'] == '':\n",
    "        human_entrez = Request_Utilis.get_entrez_id(row['GENE SYMBOL'])\n",
    "        sec_recon_dc.at[i, 'ENTREZID'] = human_entrez\n",
    "\n",
    "if not sec_recon_dc.equals(sec_recon):\n",
    "    gsheet_file.update_google_sheet(sec_recon_sheet, sec_recon_dc)\n",
    "    print(\"Google Sheet updated.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "127550\n",
      "Homo sapiens A3GALT2 alpha 1,3-galactosyltransferase 2 ['IGB3S', 'IGBS3S', 'A3GALT2P'] ENSG00000184389 [('NM_001080438', 'NP_001073907', ['U3KPV4']), ('NM_001080438', 'NP_001073907', ['U3KPV4'])]\n",
      "Mus musculus A3galt2 alpha 1,3-galactosyltransferase 2 ['iGb3', 'Gm433', 'iGb3S'] ENSMUSG00000028794 [('NM_001009819', 'NP_001009819', ['B2KGX5', 'B2KGX6', 'M9MMK2', 'Q3V1N9'])]\n",
      "Cricetulus griseus A3galt2 alpha 1,3-galactosyltransferase 2 [] ENSCGRG00015008603 [('XM_016974743', 'XP_016830232', ['A0A8C2MUL3', 'A0A8C2QMY7']), ('XM_027398983', 'XP_027254784', ['A0A8C2MUL3', 'A0A8C2QMY7'])]\n",
      "---------------------------------\n",
      "53947\n",
      "Homo sapiens A4GALT alpha 1,4-galactosyltransferase (P1PK blood group) ['P1', 'PK', 'Gb3S', 'P(k)', 'P1PK', 'A14GALT', 'A4GALT1'] ENSG00000128274 [('XM_005261648', 'XP_005261705', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('NM_001318038', 'NP_001304967', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('XM_047441412', 'XP_047297368', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('XM_005261647', 'XP_005261704', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('NM_017436', 'NP_059132', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('XM_005261644', 'XP_005261701', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('XM_017028831', 'XP_016884320', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('XM_005261646', 'XP_005261703', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('XM_011530233', 'XP_011528535', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('XM_006724266', 'XP_006724329', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('XM_006724265', 'XP_006724328', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('XM_054325716', 'XP_054181691', ['Q5D079', 'Q7Z7C6']), ('NM_001318038', 'NP_001304967', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('XM_054325710', 'XP_054181685', ['Q5D079', 'Q7Z7C6']), ('XM_054325717', 'XP_054181692', ['Q5D079', 'Q7Z7C6']), ('NM_017436', 'NP_059132', ['A0A0S2Z5J1', 'A9XEA7', 'B0L3P6', 'B2R7C4', 'I6S6F0', 'Q540I5', 'Q540I6', 'Q6RJV9', 'Q6RJW0', 'Q6RJW1', 'Q7Z2I1', 'Q9NPC4', 'Q9P1X5']), ('XM_054325714', 'XP_054181689', ['Q5D079', 'Q7Z7C6']), ('XM_054325712', 'XP_054181687', ['Q5D079', 'Q7Z7C6']), ('XM_054325711', 'XP_054181686', ['Q5D079', 'Q7Z7C6']), ('XM_054325713', 'XP_054181688', ['Q5D079', 'Q7Z7C6']), ('XM_054325709', 'XP_054181684', ['Q5D079', 'Q7Z7C6']), ('XM_054325715', 'XP_054181690', ['Q5D079', 'Q7Z7C6']), ('XM_054325708', 'XP_054181683', ['Q5D079', 'Q7Z7C6'])]\n",
      "Mus musculus A4galt alpha 1,4-galactosyltransferase [] ENSMUSG00000047878 [('NM_001004150', 'NP_001004150', ['Q0R0H6', 'Q3TRS1']), ('NM_001370647', 'NP_001357576', ['Q0R0H6', 'Q3TRS1']), ('NM_001170954', 'NP_001164425', ['Q0R0H6', 'Q3TRS1']), ('NM_001416098', 'NP_001403027', ['Q0R0H6', 'Q3TRS1']), ('XM_036159402', 'XP_036015295', ['Q3TRS1', 'Q3UF00', 'Q67BJ4'])]\n",
      "Cricetulus griseus A4galt alpha 1,4-galactosyltransferase (P1PK blood group) [] ENSCGRG00015041736 [('XM_035459542', 'XP_035315433', ['A0A8C2LT00', 'G3HXT0']), ('XM_035439190', 'XP_035295081', ['A0A8C2LT00', 'G3HXT0'])]\n",
      "---------------------------------\n",
      "51146\n",
      "Homo sapiens A4GNT alpha-1,4-N-acetylglucosaminyltransferase ['alpha4GnT'] ENSG00000118017 [('XM_017006544', 'XP_016862033', ['Q0VDK1', 'Q0VDK2', 'Q9UNA3']), ('XM_017006543', 'XP_016862032', ['Q0VDK1', 'Q0VDK2', 'Q9UNA3']), ('NM_016161', 'NP_057245', ['Q0VDK1', 'Q0VDK2', 'Q9UNA3']), ('XM_054346723', 'XP_054202698', ['Q0VDK1', 'Q0VDK2', 'Q9UNA3']), ('XM_054346722', 'XP_054202697', ['Q0VDK1', 'Q0VDK2', 'Q9UNA3']), ('NM_016161', 'NP_057245', ['Q0VDK1', 'Q0VDK2', 'Q9UNA3'])]\n",
      "Mus musculus A4gnt alpha-1,4-N-acetylglucosaminyltransferase ['Gm798', 'Alpha4gnt'] ENSMUSG00000037953 [('XM_006511246', 'XP_006511309', ['Q14BT6']), ('NM_001077424', 'NP_001070892', ['Q14BT6']), ('XM_006511244', 'XP_006511307', [])]\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'Gene-ref_db'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 17\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28mprint\u001b[39m(org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products)\n\u001b[1;32m     16\u001b[0m \u001b[38;5;66;03m# print CHO Info\u001b[39;00m\n\u001b[0;32m---> 17\u001b[0m org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products \u001b[38;5;241m=\u001b[39m \u001b[43mRequest_Utilis\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mGene_Info_from_EntrezID\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcho_ortholog_EntrezID\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;28mprint\u001b[39m(org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products)\n\u001b[1;32m     19\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m---------------------------------\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Documents/GitHub/Secretory-Pathway-Reconstruction/Gene_Info/Request_Utilis.py:139\u001b[0m, in \u001b[0;36mGene_Info_from_EntrezID\u001b[0;34m(EntrezID)\u001b[0m\n\u001b[1;32m    136\u001b[0m     gene_name \u001b[38;5;241m=\u001b[39m gene_record[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEntrezgene_locus\u001b[39m\u001b[38;5;124m'\u001b[39m][\u001b[38;5;241m0\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGene-commentary_products\u001b[39m\u001b[38;5;124m'\u001b[39m][\u001b[38;5;241m0\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGene-commentary_label\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m    138\u001b[0m \u001b[38;5;66;03m# Get Ensemble ID\u001b[39;00m\n\u001b[0;32m--> 139\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m gene_db_refs \u001b[38;5;129;01min\u001b[39;00m \u001b[43mgene_record\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mEntrezgene_gene\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mGene-ref\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mGene-ref_db\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m:\n\u001b[1;32m    140\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m gene_db_refs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDbtag_db\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEnsembl\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m    141\u001b[0m         gene_ensemble \u001b[38;5;241m=\u001b[39m (gene_db_refs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDbtag_tag\u001b[39m\u001b[38;5;124m'\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mObject-id\u001b[39m\u001b[38;5;124m'\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mObject-id_str\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Gene-ref_db'"
     ]
    }
   ],
   "source": [
    "for gene_EntrezID in sec_recon['ENTREZID']:\n",
    "\n",
    "    # Get CHO and Mouse Orthologs\n",
    "    mouse_ortholog_EntrezID = Request_Utilis.get_gene_ids(gene_EntrezID, '10090')\n",
    "    cho_ortholog_EntrezID = Request_Utilis.get_gene_ids(gene_EntrezID, '10029')\n",
    "    \n",
    "    # print Human Info\n",
    "    print(gene_EntrezID)\n",
    "    org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products = Request_Utilis.Gene_Info_from_EntrezID(gene_EntrezID)\n",
    "    print(org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products)\n",
    "    \n",
    "    # print Mouse Info\n",
    "    org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products = Request_Utilis.Gene_Info_from_EntrezID(mouse_ortholog_EntrezID)\n",
    "    print(org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products)\n",
    "    \n",
    "    # print CHO Info\n",
    "    org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products = Request_Utilis.Gene_Info_from_EntrezID(cho_ortholog_EntrezID)\n",
    "    print(org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products)\n",
    "    print(\"---------------------------------\")\n",
    "\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cho_ortholog_EntrezID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "org, gene_symbol, gene_name, gene_synonyms, gene_ensemble, gene_products = Request_Utilis.Gene_Info_from_EntrezID(cho_ortholog_EntrezID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gene_ID(gene_EntrezID):\n",
    "    gene_fetch_tries = 0\n",
    "    gene_fetch_found = True\n",
    "    record = []\n",
    "    while gene_fetch_tries < 3 and gene_fetch_found:\n",
    "        try:\n",
    "            search_results = Entrez.esearch(db=\"gene\", term = gene_EntrezID)            \n",
    "            record = Entrez.read(search_results)\n",
    "            gene_fetch_found = False\n",
    "        except Exception as e:\n",
    "            time.sleep(3)\n",
    "            gene_fetch_tries += 1\n",
    "    if gene_fetch_tries > 2:        \n",
    "        print(f\"EntrezID of {gene_EntrezID} could not be read; An error occurred\")\n",
    "\n",
    "    if int(record[\"Count\"]) > 0:\n",
    "\n",
    "        # Get the Gene ID of the first result\n",
    "        gene_id = record[\"IdList\"][0]\n",
    "\n",
    "        # Fetch the gene information, maximum 3 tries\n",
    "        fetch_tries = 0\n",
    "        fetch_found = True\n",
    "        while fetch_tries < 3 and fetch_found:\n",
    "            gene_fetch_tries = 0\n",
    "            gene_fetch_found = True\n",
    "            gene_info = []\n",
    "            while gene_fetch_tries < 3 and gene_fetch_found:\n",
    "                try:\n",
    "                    gene_info = Entrez.efetch(db=\"gene\", id=gene_id, retmode=\"xml\")    \n",
    "                    gene_fetch_found = False\n",
    "                except Exception as e:\n",
    "                    time.sleep(3)\n",
    "                    gene_fetch_tries += 1\n",
    "                    print(f\"Gene ID {gene_id}; An error occurred: {e}\")\n",
    "            if gene_fetch_tries > 2:\n",
    "                print(f\"{gene_EntrezID}. Gene ID {gene_id}; An error occurred\")\n",
    "\n",
    "            if gene_info == []:\n",
    "                KeyError(\"sda\")\n",
    "            \n",
    "            try:\n",
    "                gene_record = Entrez.read(gene_info)\n",
    "                fetch_found = False\n",
    "            except Exception as e:\n",
    "                time.sleep(3)\n",
    "                fetch_tries += 1                \n",
    "    return gene_record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EntrezID = 127550\n",
    "gene_record = read_gene_ID(int(EntrezID))\n",
    "org = gene_record[0]['Entrezgene_source']['BioSource']['BioSource_org']['Org-ref']['Org-ref_taxname']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_products = []\n",
    "org = gene_record[0]['Entrezgene_source']['BioSource']['BioSource_org']['Org-ref']['Org-ref_taxname']\n",
    "if 'Gene-ref_syn' in gene_record[0]['Entrezgene_gene']['Gene-ref']:\n",
    "    gene_synonyms = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_syn']\n",
    "else:\n",
    "    gene_synonyms = []\n",
    "\n",
    "# Get the gene name and gene NCBI ID\n",
    "if 'Gene-ref_locus' in gene_record[0]['Entrezgene_gene']['Gene-ref']:\n",
    "    gene_symbol = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_locus']                        \n",
    "else:\n",
    "    gene_symbol = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_locus-tag']\n",
    "\n",
    "# Get the Gene description\n",
    "if 'Gene-ref_desc' in gene_record[0]['Entrezgene_gene']['Gene-ref']:\n",
    "    gene_name = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_desc']\n",
    "else:\n",
    "    gene_name = gene_record[0]['Entrezgene_locus'][0]['Gene-commentary_products'][0]['Gene-commentary_label']\n",
    "\n",
    "# Get Ensemble ID\n",
    "for gene_db_refs in gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_db']:\n",
    "    if gene_db_refs['Dbtag_db'] == 'Ensembl':\n",
    "        gene_ensemble = (gene_db_refs['Dbtag_tag']['Object-id']['Object-id_str'])\n",
    "\n",
    "#print(\"Organism: \", org, \"|Entrez ID: \", EntrezID, \"|Gene Symbol: \", gene_symbol, \"|Gene Name: \", gene_name, \"|Gene Synonyms: \", gene_synonyms, \"| Gene Ensemble: \", gene_ensemble)\n",
    "\n",
    "for assembly_specific_info in gene_record[0]['Entrezgene_locus']:  \n",
    "    if 'Gene-commentary_heading' in assembly_specific_info:\n",
    "        for assembly_specific_transcript in assembly_specific_info['Gene-commentary_products']:\n",
    "            if 'Gene-commentary_products' in assembly_specific_transcript:\n",
    "                transcript_sequence_ID = assembly_specific_transcript['Gene-commentary_accession']\n",
    "\n",
    "                for Entrez_Comments in gene_record[0]['Entrezgene_comments']:\n",
    "                    if 'Gene-commentary_comment' in Entrez_Comments:\n",
    "                        for comments in Entrez_Comments['Gene-commentary_comment']:\n",
    "                            if 'Gene-commentary_products' in comments:\n",
    "                                for product_per_assembly in comments['Gene-commentary_products']:                                \n",
    "                                    if product_per_assembly['Gene-commentary_heading'] == 'mRNA Sequence':\n",
    "                                        mRNA = product_per_assembly['Gene-commentary_accession']\n",
    "                                        if mRNA == transcript_sequence_ID:  \n",
    "                                            for protein_per_transcript_per_assemlby in product_per_assembly['Gene-commentary_products']:\n",
    "                                                protein_sequence_ID = protein_per_transcript_per_assemlby['Gene-commentary_accession']\n",
    "                                                \n",
    "                                                uniprotID_list = []\n",
    "                                                for protein_comments in protein_per_transcript_per_assemlby['Gene-commentary_comment']:\n",
    "                                                    if protein_comments['Gene-commentary_heading'] == 'UniProtKB':\n",
    "                                                        for uniprotID_protein_per_assembly in protein_comments['Gene-commentary_comment'][0]['Gene-commentary_source']:\n",
    "                                                            protein_uniprot_id = uniprotID_protein_per_assembly['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_str']\n",
    "                                                            uniprotID_list.append(protein_uniprot_id)\n",
    "                                                gene_products.append(((transcript_sequence_ID, protein_sequence_ID, uniprotID_list)))                                       \n",
    "                                    else:   \n",
    "                                        if 'Gene-commentary_products' in product_per_assembly:\n",
    "                                            for transcript_per_assembly in product_per_assembly['Gene-commentary_products']:      \n",
    "                                                mRNA = transcript_per_assembly['Gene-commentary_accession']          \n",
    "                                                if mRNA == transcript_sequence_ID:                                                  \n",
    "                                                    for protein_per_transcript_per_assemlby in transcript_per_assembly['Gene-commentary_products']:                                                                \n",
    "                                                        protein_sequence_ID = protein_per_transcript_per_assemlby['Gene-commentary_accession']\n",
    "                                                        uniprotID_list = []\n",
    "                                                        if 'Gene-commentary_comment' in protein_per_transcript_per_assemlby:\n",
    "                                                            for protein_comments in protein_per_transcript_per_assemlby['Gene-commentary_comment']:                                                         \n",
    "                                                                if protein_comments['Gene-commentary_heading'] == 'UniProtKB':\n",
    "                                                                    for uniprotID_protein_per_assembly in protein_comments['Gene-commentary_comment'][0]['Gene-commentary_source']:\n",
    "                                                                        protein_uniprot_id = uniprotID_protein_per_assembly['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_str']\n",
    "                                                                        uniprotID_list.append(protein_uniprot_id)\n",
    "                                                        else:\n",
    "                                                            uniprotID_list = []\n",
    "                                                        gene_products.append(((transcript_sequence_ID, protein_sequence_ID, uniprotID_list)))\n",
    "print(org)\n",
    "print(EntrezID)\n",
    "print(gene_symbol)\n",
    "print(gene_name)\n",
    "print(gene_synonyms)\n",
    "print(gene_ensemble)\n",
    "print(gene_products)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_per_transcript_per_assemlby"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
